src/main/scala/lms/collection/PointerOps.scala

```diff
@@ -0,0 +1,48 @@
+package lms.collection
+
+import lms.core._
+import lms.util._
+import lms.core.stub._
+import lms.core.Backend._
+import lms.core.virtualize
+import lms.core.utils.time
+import lms.macros.SourceContext
+import scala.collection.mutable.HashMap
+
+trait PointerOps { b: Base =>
+  object Pointer {
+    def apply[A: Manifest](x: Var[A])(implicit pos: SourceContext) = {
+      val a = Wrap[Pointer[A]](Adapter.g.reflect("pointer-new", UnwrapV(x)))
+      cache += ((Unwrap(a), UnwrapV(x)))
+      a
+    }
+    def apply[A:Manifest](x: Rep[Array[A]])(implicit pos: SourceContext) = {
+      val a = Wrap[Pointer[A]](Adapter.g.reflect("pointer-from-array", Unwrap(x)))
+      cache += ((Unwrap(a), Unwrap(x)))
+      a
+    }
+    def deref[A: Manifest](x: Rep[Pointer[A]]) = {
+      Wrap[A](cache.getOrElse(Unwrap(x), ???))
+    }
+    private[this] val cache = HashMap[lms.core.Backend.Exp, lms.core.Backend.Exp]()
+  }
+
+  abstract class Pointer[T] extends Manifest[T]
```

Comment:
what does it mean by extending Manifest[T]?

Reply:
I am not very sure 😂 . 

I thought since we do `Wrap[+A:Manifest]`, then all Rep[A] must have A <: Manifest, so I thought Pointer might also needs to extends Manifest. But just tried and found no compilation error, so looks like it is not necessary.

Reply:
ok, but there needs a Manifest for T, right? like
`abstract class Pointer[T : Manifest]`

Reply:
Em. I agree with you. Strangely, it has not given me any compilation errors. Will fix.